libs/providers/flagd/src/lib/service/grpc/grpc-service.ts

```diff
@@ -153,42 +145,33 @@ export class GRPCService implements Service {
     return this.resolve(this._client.resolveObject, flagKey, context, logger);
   }
 
-  private connectStream(
-    connectCallback: () => void,
+  private listen(
+    reconnectCallback: () => void,
     changedCallback: (flagsChanged: string[]) => void,
     disconnectCallback: () => void,
-  ): Promise<void> {
-    return new Promise((resolve, reject) => {
-      this.logger?.debug(`${FlagdProvider.name}: connecting stream, attempt ${this._streamConnectAttempt}...`);
-      const stream = this._client.eventStream({}, {});
-      stream.on('error', (err: ServiceError | undefined) => {
-        if (err?.code === status.CANCELLED) {
-          this.logger?.debug(`${FlagdProvider.name}: stream cancelled, will not be re-established`);
+    resolveConnect?: () => void,
+    rejectConnect?: (reason: Error) => void,
+  ) {
+    this.logger?.debug(`${FlagdProvider.name}: connecting stream...`);
+    const stream = this._client.eventStream({}, {});
+    stream.on('error', (err: Error) => {
+      rejectConnect?.(err);
+      this.handleError(reconnectCallback, changedCallback, disconnectCallback);
+    });
+    stream.on('data', (message) => {
+      if (message.type === EVENT_PROVIDER_READY) {
+        this.logger?.debug(`${FlagdProvider.name}: streaming connection established with flagd`);
+        // if resolveConnect is undefined, this is a reconnection; we only want to fire the reconnect callback in that case
+        if (resolveConnect) {
+          resolveConnect();
         } else {
-          this.handleError(reject, connectCallback, changedCallback, disconnectCallback);
-        }
-      });
-      stream.on('close', () => {
-        this.handleClose();
-      });
-      stream.on('data', (message) => {
-        if (message.type === EVENT_PROVIDER_READY) {
-          this.handleProviderReady(resolve, connectCallback);
-        } else if (message.type === EVENT_CONFIGURATION_CHANGE) {
-          this.handleFlagsChanged(message, changedCallback);
+          reconnectCallback();
```

Comment:
When listen is called the first time (with a resolve() callback) we don't want to emit our own READY event, when we are reconnecting, we do. This fixes some duplicate ready events.